File: src/main/java/com/teamwizardry/wizardry/common/block/entity/craftingplate/BlockCraftingPlate.java
package com.teamwizardry.wizardry.common.block.entity.craftingplate;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import com.teamwizardry.wizardry.common.block.IManaNode;
import com.teamwizardry.wizardry.common.init.ModBlocks;

import net.minecraft.block.Block;
import net.minecraft.block.BlockRenderType;
import net.minecraft.block.BlockState;
import net.minecraft.block.BlockWithEntity;
import net.minecraft.block.ShapeContext;
import net.minecraft.block.Waterloggable;
import net.minecraft.block.entity.BlockEntity;
import net.minecraft.block.entity.BlockEntityTicker;
import net.minecraft.block.entity.BlockEntityType;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.fluid.FluidState;
import net.minecraft.fluid.Fluids;
import net.minecraft.item.ItemPlacementContext;
import net.minecraft.item.ItemStack;
import net.minecraft.state.StateManager;
import net.minecraft.state.property.BooleanProperty;
import net.minecraft.state.property.Properties;
import net.minecraft.util.ActionResult;
import net.minecraft.util.Hand;
import net.minecraft.util.ItemScatterer;
import net.minecraft.util.hit.BlockHitResult;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import net.minecraft.util.shape.VoxelShape;
import net.minecraft.world.BlockView;
import net.minecraft.world.World;
import net.minecraft.world.WorldAccess;

public class BlockCraftingPlate extends BlockWithEntity implements Waterloggable, IManaNode
{
    public static final BooleanProperty WATERLOGGED = Properties.WATERLOGGED;
    private static final VoxelShape SHAPE = Block.createCuboidShape(0.0D, 0.0D, 0.0D, 16.0D, 13.0D, 16.0D);

    public BlockCraftingPlate(Settings settings)
    {
        super(settings);
        this.setDefaultState(this.stateManager.getDefaultState().with(WATERLOGGED, Boolean.FALSE));
    }

    @Override public BlockRenderType getRenderType(BlockState state) { return BlockRenderType.MODEL; }
    
    @Override public VoxelShape getOutlineShape(BlockState state, BlockView view, BlockPos pos, ShapeContext context) { return SHAPE; }

    @Override
    public <T extends BlockEntity> BlockEntityTicker<T> getTicker(World world, BlockState state, BlockEntityType<T> type) {
        if (world.isClient) return checkType(type, ModBlocks.craftingPlateEntity, (w,p,s,e) -> BlockCraftingPlateEntity.clientTick(w,e));
        else return checkType(type, ModBlocks.craftingPlateEntity, (w,p,s,e) -> BlockCraftingPlateEntity.serverTick(world, e));
    }
    
    @Override
    public @NotNull ActionResult onUse(BlockState state, World world, BlockPos pos, PlayerEntity player, Hand hand, BlockHitResult hit)
    {
        if (!world.isClient)
        {
            if (hand == Hand.MAIN_HAND)
            {
                ItemStack heldItem = player.getStackInHand(hand);
                BlockEntity blockEntity = world.getBlockEntity(pos);
                if (blockEntity instanceof BlockCraftingPlateEntity)
                {
                    BlockCraftingPlateEntity plate = (BlockCraftingPlateEntity) blockEntity;
                    if (heldItem.isEmpty())
                        player.setStackInHand(hand, plate.removeItem());
                    else
                        plate.addItem(heldItem.split(1));
                }
            }
            return ActionResult.SUCCESS;
        }
        return super.onUse(state, world, pos, player, hand, hit);
    }

    @Override
    public void onEntityCollision(BlockState state, World world, BlockPos pos, Entity entity)
    {
        BlockEntity blockEntity = world.getBlockEntity(pos);
        if (blockEntity instanceof BlockCraftingPlateEntity)
            ((BlockCraftingPlateEntity) blockEntity).onEntityCollision(entity);
    }

    @Nullable @Override public BlockCraftingPlateEntity createBlockEntity(BlockPos pos, BlockState state) { return new BlockCraftingPlateEntity(pos, state); }

    @Override
    public void onStateReplaced(BlockState state, World world, BlockPos pos, BlockState newState, boolean moved)
    {
        BlockEntity blockEntity = world.getBlockEntity(pos);
        if (blockEntity instanceof BlockCraftingPlateEntity)
            ItemScatterer.spawn(world, pos, ((BlockCraftingPlateEntity) blockEntity).getInventory());
        super.onStateReplaced(state, world, pos, newState, moved);
    }

    @Nullable
    @Override
    public BlockState getPlacementState(ItemPlacementContext context)
    {
        return this.getDefaultState().with(WATERLOGGED, context.getWorld().getFluidState(context.getBlockPos()).getFluid() == Fluids.WATER);
    }

    @Override
    public BlockState getStateForNeighborUpdate(BlockState state, Direction direction, BlockState newState, WorldAccess world, BlockPos pos, BlockPos posFrom)
    {
        if (state.get(WATERLOGGED))
            world.getFluidTickScheduler().schedule(pos, Fluids.WATER, Fluids.WATER.getTickRate(world));
        return super.getStateForNeighborUpdate(state, direction, newState, world, pos, posFrom);
    }

    @Override public FluidState getFluidState(BlockState state) { return state.get(WATERLOGGED) ? Fluids.WATER.getStill(false) : super.getFluidState(state); }
    
    @Override protected void appendProperties(StateManager.Builder<Block, BlockState> builder) { builder.add(WATERLOGGED); }

    @Override public ManaNodeType getManaNodeType() { return ManaNodeType.SINK; }
}

File: src/main/java/com/teamwizardry/wizardry/common/block/entity/craftingplate/BlockCraftingPlate.kt
package com.teamwizardry.wizardry.common.block.entity.craftingplate

import com.teamwizardry.wizardry.common.init.ModBlocks
import net.minecraft.block.*
import net.minecraft.block.entity.BlockEntityType
import net.minecraft.entity.Entity
import net.minecraft.item.ItemStack
import net.minecraft.state.property.Properties
import net.minecraft.util.Hand
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.world.BlockView

class BlockCraftingPlate(settings: Settings?) : BlockWithEntity(settings), Waterloggable, IManaNode {
    override fun getRenderType(state: BlockState): BlockRenderType {
        return BlockRenderType.MODEL
    }

    override fun getOutlineShape(state: BlockState, view: BlockView, pos: BlockPos, context: ShapeContext): VoxelShape {
        return SHAPE
    }

    override fun <T : BlockEntity?> getTicker(
        world: World,
        state: BlockState,
        type: BlockEntityType<T>
    ): BlockEntityTicker<T>? {
        return if (world.isClient) checkType(
            type,
            ModBlocks.craftingPlateEntity,
            BlockEntityTicker<BlockCraftingPlateEntity> { w: World?, p: BlockPos?, s: BlockState?, e: BlockCraftingPlateEntity? ->
                BlockCraftingPlateEntity.clientTick(
                    w,
                    e
                )
            }) else checkType(
            type,
            ModBlocks.craftingPlateEntity,
            BlockEntityTicker<BlockCraftingPlateEntity> { w: World?, p: BlockPos?, s: BlockState?, e: BlockCraftingPlateEntity ->
                BlockCraftingPlateEntity.serverTick(
                    world,
                    e
                )
            })
    }

    override fun onUse(
        state: BlockState,
        world: World,
        pos: BlockPos,
        player: PlayerEntity,
        hand: Hand,
        hit: BlockHitResult
    ): ActionResult {
        if (!world.isClient) {
            if (hand == Hand.MAIN_HAND) {
                val heldItem: ItemStack = player.getStackInHand(hand)
                val blockEntity: BlockEntity = world.getBlockEntity(pos)
                if (blockEntity is BlockCraftingPlateEntity) {
                    val plate = blockEntity as BlockCraftingPlateEntity
                    if (heldItem.isEmpty) player.setStackInHand(
                        hand,
                        plate.removeItem()
                    ) else plate.addItem(heldItem.split(1))
                }
            }
            return ActionResult.SUCCESS
        }
        return super.onUse(state, world, pos, player, hand, hit)
    }

    override fun onEntityCollision(state: BlockState, world: World, pos: BlockPos, entity: Entity) {
        val blockEntity: BlockEntity = world.getBlockEntity(pos)
        if (blockEntity is BlockCraftingPlateEntity) (blockEntity as BlockCraftingPlateEntity).onEntityCollision(entity)
    }

    override fun createBlockEntity(pos: BlockPos, state: BlockState): BlockCraftingPlateEntity {
        return BlockCraftingPlateEntity(pos, state)
    }

    override fun onStateReplaced(state: BlockState, world: World, pos: BlockPos, newState: BlockState, moved: Boolean) {
        val blockEntity: BlockEntity = world.getBlockEntity(pos)
        if (blockEntity is BlockCraftingPlateEntity) ItemScatterer.spawn(
            world,
            pos,
            (blockEntity as BlockCraftingPlateEntity).inventory
        )
        super.onStateReplaced(state, world, pos, newState, moved)
    }

    override fun getPlacementState(context: ItemPlacementContext): BlockState? {
        return this.defaultState.with(
            WATERLOGGED,
            context.getWorld().getFluidState(context.getBlockPos()).getFluid() === Fluids.WATER
        )
    }

    override fun getStateForNeighborUpdate(
        state: BlockState,
        direction: Direction,
        newState: BlockState,
        world: WorldAccess,
        pos: BlockPos,
        posFrom: BlockPos
    ): BlockState {
        if (state.get(WATERLOGGED)) world.getFluidTickScheduler()
            .schedule(pos, Fluids.WATER, Fluids.WATER.getTickRate(world))
        return super.getStateForNeighborUpdate(state, direction, newState, world, pos, posFrom)
    }

    override fun getFluidState(state: BlockState): FluidState {
        return if (state.get(WATERLOGGED)) Fluids.WATER.getStill(false) else super.getFluidState(state)
    }

    override fun appendProperties(builder: StateManager.Builder<Block, BlockState>) {
        builder.add(WATERLOGGED)
    }

    val manaNodeType: ManaNodeType
        get() = ManaNodeType.SINK

    companion object {
        val WATERLOGGED: BooleanProperty = Properties.WATERLOGGED
        private val SHAPE: VoxelShape = createCuboidShape(0.0, 0.0, 0.0, 16.0, 13.0, 16.0)
    }

    init {
        this.defaultState = this.stateManager.defaultState.with(WATERLOGGED, java.lang.Boolean.FALSE)
    }
}